### DISCUSSION TITLE
Cache StringNames and NodePaths when casting from string in C#

### DISCUSSION THREAD

**Discussion by Joy-less** [Redot-Engine/redot-proposals#50] [Ideas]
I previously discussed this in detail in the Godot Proposals repository: https://github.com/godotengine/godot-proposals/discussions/10826

The TL;DR is that StringNames and NodePaths create insidious [garbage collection performance spikes](https://www.reddit.com/r/godot/comments/17tqipk) in C#, due to allocation of new objects inside implicit casts.

```cs
Input.IsActionPressed("action"); // Allocated a StringName.
GetNode("path/to/node"); // Allocated a NodePath.
```

This is my current workaround:
```cs
global using static RecycleExtensions;

using ZiggyCreatures.Caching.Fusion;

public static class RecycleExtensions {
    private static readonly FusionCache StringNameCache = new(new FusionCacheOptions());
    private static readonly FusionCache NodePathCache = new(new FusionCacheOptions());

    /// <summary>
    /// Converts the string to a StringName, recycling if possible.
    /// </summary>
    public static StringName StringName(string String) {
        return StringNameCache.GetOrSet(String, _ => (StringName)String);
    }
    /// <summary>
    /// Converts the string to a NodePath, recycling if possible.
    /// </summary>
    public static NodePath NodePath(string String) {
        return NodePathCache.GetOrSet(String, _ => (NodePath)String);
    }
}
```
```cs
StringName MyStringName = StringName("test");
NodePath MyNodePath = NodePath("test");
```

The solution would be one of the following:
1. Remove StringNames and NodePaths, since they don't add much value.
2. Remove StringNames and NodePaths from the C# API.
3. Put the caching logic from the workaround inside the implicit casts to StringName/NodePath (simplest solution).

**Comment by Arctis-Fireblight** [comment]
Hey @Joy-less ,
I went ahead and reviewed your PR https://github.com/godotengine/godot/pull/100452
It looks like a pretty solid implementation, and I am interested in merging it into Redot. Would you still be willing to submit a PR for this for inclusion in Redot?

It is a little late in the dev cycle to get it included in Redot 26.1 since I will be cutting a stable release this week most likely, but I would love to include this in 26.2!

**Reply by Joy-less** [reply]
Hi @Arctis-Fireblight
I sent the pull request here: https://github.com/Redot-Engine/redot-engine/pull/1175

Hopefully it can be merged into Godot Engine at some point as well :)
